### ISSUE TITLE
Autotune crashed my wing

### ISSUE THREAD

**Issue by shellixyz** [iNavFlight/inav#2036]
I was trying autotune on my wing and it made it lose control. I couldn't get it to level again and it was spinning on the roll axis. I tried two times with the same result. I was flying in angle mode and did not act fast enough to manage to regain control and crashed nose first into the ground. I attached the log of the second attempt.
[LOG00011.TXT](https://github.com/iNavFlight/inav/files/1252148/LOG00011.TXT)

**Comment by digitalentity** [comment]
Are you sure your wing can do 1.1 barrel rolls per second? You log specifies that you have configured max roll rate of 400 deg/s - that's a bit too much for most airplanes. Autotune is not designed to handle high rates for aerobatic flights.

UPDATE: Would also be good to see your configuration. It seems that max_angle_inclination is about 70-75 deg which is also close to the point where leveling controller is loosing stability.

**Comment by shellixyz** [comment]
I'm pretty sure it can, it is fast, but it is difficult to measure without video. I was not rolling when it happened. Each times it was when I pitched hard up/down. I flew it hard since without autotune and I had no problem. Yes max roll angle is high but the problem did not happen when rolling but when pitching up/down. I attached the config.

EDIT: It started acting up at 2:31.9. In the log you can see I pitched hard down and then nothing it just went down I couldn't do anything. I should have switched back to passthrough but there wasn't much time to do so.

[config_20170826.txt](https://github.com/iNavFlight/inav/files/1253934/config_20170826.txt)

**Comment by shellixyz** [comment]
At 2:33.5 you can see the roll/pitch input is 0 and although I was in angle mode it was nose down and rolling.

**Comment by digitalentity** [comment]
Yes, 90 deg pitch is a singularity point, where roll becomes meaningless. In that case wing can't get out of steep dive because roll transitions from -180 to +180 randomly.

**Comment by shellixyz** [comment]
I understand that but I don't understand exactly what is the problem then. The max pitch angle is set to 45°. Why would it go to 90° ? The first time I tried to recover by switching to "Acro" but couldn't recover either. In any case it seems to me that autotune is very dangerous. What am I doing wrong ?

**Comment by digitalentity** [comment]
You need to understand how autotune works. When you command roll/pitch autotune analyzes your command vs wing response. If wing responds too fast and exceeds the requested rate - autotune reduces PIFF gains. If wing responds too slow - autotune increses gains. In most cases your wing responded faster than expected and autotune 

Looked at the part of log you mentioned (~2:30) - it's very interesting. When you tried to pitch up the wing instead rolled right. Firmware was trying to level it back as hard as possible, however wing didn't respond.

In the earlier part of log - this is normal behavior - gyro (red) is following PID output (yellow) with a delay:
![image](https://user-images.githubusercontent.com/11059099/29746260-b5487ec0-8b17-11e7-9a2d-a79fb36fce02.png)

However here (at uncontrolled roll) there is no correlation between PID output and wing response:
![image](https://user-images.githubusercontent.com/11059099/29746264-e9096ac6-8b17-11e7-91e1-fe85adf3429d.png)

In both cases PID output (and thus servo deflection) is of similar magnitude.

I don't think this is algorithm flaw - PID controller is working as expected, wing was not in 90deg dive but neither pitch, not roll was responding to servo commands. My best guess is a failed servo or linkage.

Also you wrote this:
> The first time I tried to recover by switching to "Acro" but couldn't recover either. 

The moment you switch out of autotune PID gains from 5 seconds ago are restored. And the fact that wing didn't respond to your commands in ACRO with good PID gains also looks like a hardware issue.

**Comment by shellixyz** [comment]
I see. I thought about it and I realized I probably made a mistake in my servo mix. I mixed 100% for pitch and roll as you can see in my configuration when it should be 50%. Do you think it could be the cause ? I don't think it is a servo or linkage issue as they were not damaged during the crash and I flew the wing without changing anything related to the elevons and it flies just fine. It probably reacted too quickly as you said and got stuck nose down because it oscillated between 100% roll in one direction and the other and couldn't apply pitch correction. I will have to test it again after changing my mix to 50% for roll and pitch but I'm still afraid it could crash.

**Comment by digitalentity** [comment]
Might be the case, however I doubt that mixer is responsible. PID controller didn't request more than 300 units servo throw - even with 100% that won't saturate the controls. I suggest that you reduce max roll angle to 50-60 deg and pitch angle to 30 deg. This should prevent edge cases. Also your bail-out mode should be passthrough, not ACRO - there you have full manual control over servos.

For initial test with 50% mix set roll D to 35-40 and pitch D to ~120-150 (according to what autotune ended up with). This is odd, however. Usually flying wing needs similar response gains on roll and pitch - might be CoG issue.

Btw, what wing is this?

**Comment by shellixyz** [comment]
Yes you are right I should have switched to passthrough but I'm not sure I would have managed to recover anyway. The first time I changed mode but had 2 switch changes to go in passthrough, not very clever I know. The second time I only had 3 seconds before impact and didn't even manage to change mode even though I changed my switches to only have one to change to go in passthrough. Next time I try I will probably go higher and with FPV to give me a chance to recover.

It is an experimental wing with a custom ABS center section and foam board wings based on the FT Mini Arrow. 980mm wing span.
